### PR TITLE
[ign to gz] C++ namespaces

### PR DESCRIPTION
**Pull request by chapulina**
# 🎉 New feature

* Part of https://github.com/ignition-tooling/release-tools/issues/698
* Subtask: https://github.com/ignition-tooling/release-tools/issues/711

## Summary

Changes:

* Change all namespaces from `ignition` to `gz`
* Add alias so that `ignition` still works for users (tick-tock's tick)
* Added a test to make sure `ignition` still works
* While at it, I added missing inline version namespaces in a couple of places

Ideally we'd issue deprecation warnings when using `ignition`. Unfortunately, C++ doesn't allow this:

```
namespace [[deprecated]] ignition = gz;
```

Clang's warning is very clear:

```
.../include/ignition/utils/config.hh:24:11: error: attributes cannot be specified on namespace alias
namespace [[deprecated]] ignition = gz;
          ^
```

So we'll need to either find a workaround, or don't issue deprecation warnings for namespaces.

In any case, we should wait until most of the stack is already using `gz` before we turn on these warnings, otherwise they would hide other warnings.

## Test it

See the added test. I also expect downstream libraries not to need any modifications because of this PR.

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial - We'll add complete migration tutorials at the end
- [x] Updated documentation (as needed)
- [x] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

### CODE REVIEW

File: include/ignition/utils/config.hh.in [hunk 1]
```diff
@@ -15,3 +15,10 @@
 #cmakedefine IGNITION_UTILS_BUILD_TYPE_PROFILE 1
 #cmakedefine IGNITION_UTILS_BUILD_TYPE_DEBUG 1
 #cmakedefine IGNITION_UTILS_BUILD_TYPE_RELEASE 1
+
```
Comment (scpeters) [inline include/ignition/utils/config.hh.in:18]: should we add header guards to this file?

**Review by scpeters — COMMENTED** [review]
> Ideally we'd issue deprecation warnings when using `ignition`. Unfortunately, C++ doesn't allow this:

I assume we will rename the headers and provide `ignition/utils/*` headers with `#pragma` deprecation warnings inside. I would suggest placing the `namespace ignition = gz;` statement inside one of these headers

File: include/ignition/utils/config.hh.in (hunk 1 above)
Comment (chapulina) [reply include/ignition/utils/config.hh.in:38]: Good call, added in [5fee873](https://github.com/ignitionrobotics/ign-utils/pull/50/commits/5fee87349cdfa442df89e87c9c13a34628c5c95b)

**Comment by chapulina** [comment]
> Can we use this trick?

Brilliant! ✨  Works like a charm. I added it commented out in [5fee873](https://github.com/ignitionrobotics/ign-utils/pull/50/commits/5fee87349cdfa442df89e87c9c13a34628c5c95b) so we don't enable warnings just yet. It didn't work with `IGN_DEPRECATED`, so I used `[[deprecated]]`.

> I assume we will rename the headers and provide ignition/utils/* headers with #pragma deprecation warnings inside. I would suggest placing the namespace ignition = gz; statement inside one of these headers

I was thinking of adding symlinks so we don't duplicate the headers. But I'm open to suggestions about a different place to put the alias. I went for `config.hh` because that's common across all libraries and including it adds little overhead

**Comment by scpeters** [comment]
> > I assume we will rename the headers and provide ignition/utils/* headers with #pragma deprecation warnings inside. I would suggest placing the namespace ignition = gz; statement inside one of these headers
> 
> I was thinking of adding symlinks so we don't duplicate the headers. But I'm open to suggestions about a different place to put the alias. I went for `config.hh` because that's common across all libraries and including it adds little overhead

I wouldn't suggest using identical headers in `ignition/utils/*`. I would have some deprecation warnings and then a `#include <gz/utils/___.hh>` statement, similar to the following:

* https://github.com/ignitionrobotics/sdformat/blob/sdf12/include/sdf/sdf_config.h

**Comment by chapulina** [comment]
> I wouldn't suggest using identical headers in ignition/utils/*. I would have some deprecation warnings and then a #include <gz/utils/___.hh> statement, similar to the following:

Ah ok. I'm not sure I fully understand though, do you think we should keep each one of the existing headers, and inside each of them, include the `gz` equivalent? Or would it be just one master `ignition` header that includes all the `gz` headers?

That may work better than symlinks because it's better for cross-platform 🤔

**Comment by methylDragon** [comment]
This was superceded by https://github.com/gazebosim/gz-utils/pull/61 because I had to do the namespace migration following the header migration.

**Comment by chapulina** [comment]
Closing in favor of #61

**Pull request closed without merging**